### PR TITLE
[libc++] Add exception guard for vector<bool>::__init_with_sentinel

### PR DESCRIPTION
**Pull request by winner245**
I think this was missing from earlier PRs.

### CODE REVIEW

File: libcxx/include/__vector/vector_bool.h
Comment (ldionne) [inline libcxx/include/__vector/vector_bool.h:401]: IIUC, this isn't fixing a bug, right? I think both versions are 100% equivalent.

If that's not the case, then what is the behaviour change? We should have a test for it.

**Review by ldionne — APPROVED** [review]
LGTM assuming this is NFC like it seems to be.

File: libcxx/include/__vector/vector_bool.h
Comment (winner245) [reply libcxx/include/__vector/vector_bool.h:401]: I think they are 100% equivalent. Currently, there are no exception tests for vector<bool> although we have exceptions tests for vector<T>. I can add one.

Comment (ldionne) [reply libcxx/include/__vector/vector_bool.h:401]: If you're able to add one, that would be amazing.

Comment (winner245) [reply libcxx/include/__vector/vector_bool.h:401]: I have added an exception test file for `vector<bool>`, which covers all constructors that might raise exceptions in various scenarios. Since the exception tests for `vector<bool>` and `vector<T>` are similar, I factored out the common code used by both tests into a header file `exception_test_helpers.h`. 

Additionally, I found some issues with the existing exceptions tests for `vector<T>`: some tests did not throw exception as intended, making them not useful, and some tests did not throw exceptions at the right points, failing to fulfill their purpose. We may need another PR to fix that.

File: libcxx/test/support/exception_test_helpers.h
Comment (ldionne) [inline libcxx/test/support/exception_test_helpers.h:1]: We have several helper classes in the test suite that attempt to provide reusable functionality like these ones. In practice, this often doesn't work really well because these classes are closely tied to the way they're used.

I spent some time trying to think about ways to make the classes below more "general purpose" so that it would make sense to actually reuse them in other tests (besides the ones for `std::vector<T>`) and I couldn't come up with something good. As it stands, I think it's unlikely that e.g. `throwing_allocator` would really be reused much outside of the `std::vector` classes (let me know if you disagree).

If you agree with the above, then I think it would be better to avoid putting these classes inside `support/` to avoid giving the impression that they are general purpose. Instead, we can put them in e.g. `libcxx/test/std/containers/sequences/vector/common.h`.

File: libcxx/test/support/exception_test_helpers.h
Comment (ldionne) [inline libcxx/test/support/exception_test_helpers.h:17]: This isn't telling the truth -- why? This would need a comment.

File: libcxx/test/support/exception_test_helpers.h
Comment (ldionne) [inline libcxx/test/support/exception_test_helpers.h:21]:
```suggestion
  explicit throwing_allocator(bool throw_on_ctor, bool throw_on_copy) : throw_on_copy_(throw_on_copy) {
```

Let's remove the defaults. When we construct, let's use `throwing_allocator alloc(/* throw_on_ctor */true, /* throw_on_copy */false);` (for example), to make this more explicit.

File: libcxx/test/support/exception_test_helpers.h
Comment (ldionne) [inline libcxx/test/support/exception_test_helpers.h:13]: Missing includes `<type_traits>`, `<cstddef>`, `<memory>` and `<cassert>`.

File: libcxx/test/support/exception_test_helpers.h
Comment (ldionne) [inline libcxx/test/support/exception_test_helpers.h:87]: Missing newline at the end of the file.

File: libcxx/test/std/containers/sequences/vector.bool/exceptions.pass.cpp
Comment (ldionne) [inline libcxx/test/std/containers/sequences/vector.bool/exceptions.pass.cpp:1]: Let's rename this file to `ctor.exceptions.pass.cpp` since we're only testing constructors.

Actually, writing this comment made me realize that we already have `libcxx/test/std/containers/sequences/vector.bool/ctor_exceptions.pass.cpp`. What's up with that test? We should merge this new test into that existing test instead. Do you have thoughts about that?

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/test/support/exception_test_helpers.h
Comment (winner245) [reply libcxx/test/support/exception_test_helpers.h:1]: Thank you for your suggestion. Yes, these classes, e.g., `throwing_allocator`, `throwing_iterator` were originally only used by the primary template `std::vector<T, Alloc>` before I added new tests for the `vector<bool>` specialization. During this PR, I noticed that we didn't have exception tests for `vector<bool>`, so I added these tests for the ctors of `vector<bool>`. Then I realized that the extracted classes such as `throwing_allocator`, `throwing_iterator` are now used by both the primary `std::vector<T, Alloc>` and the specialization `vector<bool>`. 

I agree with your that these classes do not belong to `support/` which targets more general purposes. Your suggestion `libcxx/test/std/containers/sequences/vector/common.h` sounds good to me.

**Comment by ldionne** [comment]
For information, I just merged #116449 so you might want to rebase on top of the latest `main`.

File: libcxx/test/std/containers/sequences/vector.bool/exceptions.pass.cpp
Comment (winner245) [reply libcxx/test/std/containers/sequences/vector.bool/exceptions.pass.cpp:1]: Oops, I should have noticed this file earlier—it would have simplified my PR.
 I just quickly reviewed this file `libcxx/test/std/containers/sequences/vector.bool/ctor_exceptions.pass.cpp` and found  it has similarly flaw as in the exception test for `std::vector`: the same `Allocator` class is used in `ctor_exceptions.pass.cpp` does not trigger exception at all, making some tests checking for nothing for `std::vector<bool>`. Also, this test file also provided duplicate definitions for the `Allocator` and `Iterator` classes. I believe we still need to refactor these common classes into a file such as `libcxx/test/std/containers/sequences/vector/common.h`, as you suggested earlier. I will work on this first.

File: libcxx/test/support/exception_test_helpers.h
Comment (winner245) [reply libcxx/test/support/exception_test_helpers.h:17]: This class was named `Allocator<T>` which sounds a general allocator class with no indication of whether it may throw exception. Renaming it from `Allocator` to `throwing_allocator` makes its purpose to throw exceptions more explicit.

Comment (winner245) [reply libcxx/test/support/exception_test_helpers.h]: Thank you for this comment, which motivated me to think more deeply about the issue. I just realized that I must have misunderstood the question initially. Now I believe the question pertains to the line `using is_always_equal = std::false_type;`, which appears counter-intuitive for a stateless allocator class. 

This line was in the original class `Allocator` before my refactor, my best guess is that the original author intended to use this class to test the exception safety for the constructor `vector(vector&&, const allocator_type&)` in `vector/vector.cons/exceptions.pass.cpp`: 

```cpp
try { // Throw in vector(vector&&, const allocator_type&) from type
    std::vector<ThrowingT, Allocator<ThrowingT> > vec(Allocator<ThrowingT>(false));
    int throw_after = 1;
    vec.emplace_back(throw_after);
    std::vector<ThrowingT, Allocator<ThrowingT> > vec2(std::move(vec), Allocator<ThrowingT>(false));
  } catch (int) {
  }
  check_new_delete_called();
```

This test aims to verify the exception safety for the constructor while performing an O(n) element-wise move (rather than the O(1) resource stealing), which requires the instances of the Allocator class to compare unequal. However, merely  setting `Allocator::is_always_equal::value` to false does not ensure that the instances of this class are unequal, which renders the above test actually testing nothing, as you also confirmed [here](https://github.com/llvm/llvm-project/pull/116449#pullrequestreview-2462660650). 
 

Beyond this scenario, I couldn't find any reason why we might need this member type. I believe we can safely remove it. Would you suggest I remove it in this PR or in my next one while rebasing?

**Comment by winner245** [comment]
I have done all the changes. Here is a brief summary for your information:

- The common classes for `vector` and `vector<bool>` tests are now organized in `vector/common.h`, with some necessary definition improvements and meaningful name changes: `Alloctor > throwing_allocator`, `Iterator > throwing_iterator`. 
- The test file `vector.bool/ctor_exceptions.pass.cpp` has been augmented with 5 additional tests: 
   - 3 of which covering exception tests for the copy/move-ctors with a specified allocator;
   - 2 of which covering exception tests for other ctors with/without a specified allocator. 
 
I think it is ready to merge. Once it is merged, I will proceed to rebase my other PRs.

File: libcxx/test/std/containers/sequences/vector/common.h
Comment (ldionne) [inline libcxx/test/std/containers/sequences/vector/common.h:97]:
```suggestion
#endif // TEST_STD_CONTAINERS_SEQUENCES_VECTOR_COMMON_H
```

**Review by ldionne — APPROVED** [review]
Thank you, this is great!

File: libcxx/test/support/exception_test_helpers.h
Comment (winner245) [reply libcxx/test/support/exception_test_helpers.h]: The above test has been replaced by the following test:

```cpp
try { // Throw in vector(vector&&, const allocator_type&) from type during element-wise move
  std::vector<throwing_t, test_allocator<throwing_t> > vec(test_allocator<throwing_t>(1));
  int throw_after = 10;
  throwing_t v(throw_after);
  vec.insert(vec.end(), 6, v);
  std::vector<throwing_t, test_allocator<throwing_t> > vec2(std::move(vec), test_allocator<throwing_t>(2));
} catch (int) {
}
check_new_delete_called();
```

This test now uses  `test_allocator<T>`, which checks equality between the allocator instances  based on the `data_` field. In the above test, the data fields 1 and 2 make the two allocators unequal, thus ensuring element-wise move. 

The original test use a stateless allocator type `Allocator` (now renamed to `throwing_allocator`), which is not suitable for the above purpose. Also, the fact that a stateless allocator type has a false `is_always_equal` value sounds a bit counter-intuitive to me. That's why I removed the member type in my later PR. Now `locator_traits<throwing_allocator>::is_always_equal` becomes true and this is consistence with its `operator==` which always returns true.

**Pull request merged**
